src/realm/sync/client.cpp

```diff
@@ -673,16 +677,42 @@ void SessionImpl::initiate_integrate_changesets(std::uint_fast64_t downloadable_
         }
         version_type client_version;
         if (REALM_LIKELY(!get_client().is_dry_run())) {
-            VersionInfo version_info;
-            ClientReplication& repl = access_realm(); // Throws
-            integrate_changesets(repl, m_progress, downloadable_bytes, changesets, version_info,
-                                 batch_state); // Throws
-            client_version = version_info.realm_version;
+            if (batch_state == DownloadBatchState::MoreToCome) {
+                if (!m_wrapper.m_flx_pending_bootstrap_store) {
+                    m_wrapper.m_flx_pending_bootstrap_store = std::make_unique<PendingBootstrapStore>(m_wrapper.m_db);
+                }
+                auto add_mode = (query_version == m_wrapper.m_flx_last_seen_version)
+                                    ? PendingBootstrapStore::Append
+                                    : PendingBootstrapStore::ClearFirst;
+                logger.debug("Caching download message until all bootstrap downloads are received");
+                client_version = m_wrapper.m_flx_pending_bootstrap_store->add_changeset_batch(add_mode, changesets);
+            }
+            else {
+                VersionInfo version_info;
+                ClientReplication& repl = access_realm(); // Throws
+                if (m_wrapper.m_flx_pending_bootstrap_store) {
+                    while (m_wrapper.m_flx_pending_bootstrap_store->has_pending()) {
```

Comment:
will we have the same problem if we don't get to apply all changeset batches here? I thought the idea was to apply all or nothing (i.e., concatenate all batches and have one call to integrate_changesets)